cmd/gh/main_test.go

```diff
@@ -76,3 +88,128 @@ check your internet connection or https://githubstatus.com
 		})
 	}
 }
+
+func Test_IntegrationRepoFork(t *testing.T) {
+	// Set up temp dir
+	tempDir := t.TempDir()
+	oldWd, _ := os.Getwd()
+	assert.NoError(t, os.Chdir(tempDir))
+	t.Cleanup(func() { _ = os.Chdir(oldWd) })
+
+	// Set up server to handle network requests
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
It's smart how you used proxy settings to redirects requests to a test server. The question now is, if we don't verify that gh works against the real GitHub API, does that compromise the validity of integration tests?

In my experience of maintaining fake API responses in the `hub` repo, the HTTP stubs would fall out of date with API behavior over time, or the stubs would often not be a complete representation of the full API response (headers and all) since I would write them by hand as opposed to having them recorded from real HTTP interactions. That's why I'm thinking: if we could base the stubs on recorded payloads, that would be even better!

Reply:
That may be possible and I did encounter https://github.com/dnaeon/go-vcr which looks like it would do what you are suggesting. My experience in Ruby using VCR http requests was not great though. They are indeed a more accurate depiction of a full http request compared to manually writing our own responses but they do not inherently solve the problem of falling out of date with the API as they need to be manually re-recorded or set to re-record at a given time interval. The re-recording of the VCRs is where I ran into a lot of pain in the past and I can envision it being pretty awful here as well. 

Some initial VCR problems that I anticipate are:
* Time values in requests, for example with repo fork response if the `created_at` time of the repo is not within a couple minutes of the current time we give the user different output saying the repo has already been forked. These values are easy to control in hand written responses.
* Re-recording VCRs requires an authenticated user.
* Re-recording VCRs requires a clean state, for example forking a repo for recording the first VCR will work but trying to fork again when re-recording the VCR will fail because the fork already exists and require manual cleanup before being able to re-record. 

Overall I don't think testing against the real GitHub API adds any more validity to the integration tests. If the API changes underneath us and breaks some functionality the only way we would catch it is if we made actual requests to the API, neither VCR or hand written responses would solve that issue. During re-recording we _might_ see that the API has changed but it is much more likely that our users would report broken functionality before then.

TL;DR: I am opposed to recorded responses because they take a lot of effort to maintain, are not easily configurable, and don't buy us much except that they are full http responses.

Reply:
All the cons you listed make sense; thank you for weighing in.